
# 🎯 Cohen's Kappa Nedir?

Cohen's Kappa, iki değerlendirici veya iki sınıflandırıcı arasındaki **uyumu** ölçmek için kullanılan bir metriktir.

Sadece yüzeysel uyumu (yani, yüzde kaç aynı sonucu verdiler) değil, **rastgele aynı tahmin etme olasılığını da dikkate alır**.

---

## 📍 Nerede Kullanılır?

- Tıp alanında iki doktorun aynı teşhisi koyma oranını ölçmek
- Veri etiketleme sürecinde etiketleyiciler arasındaki uyumu kontrol etmek
- Bir modelin insan etiketiyle ne kadar uyumlu olduğunu görmek

---

## ⚖️ Avantajları

✅ Rastgele uyumu hesaba katar  
✅ Kategorik (sınıflandırma) verilerde anlamlı

---

## ⚠️ Dezavantajları

❌ Sadece **iki değerlendirici** için uygundur  
❌ Dengesiz veri setlerinde yorumlamak zor olabilir

---

## 💬 Değer Aralığı

- **1.0**: Tam uyum
- **0.0**: Rastgele uyum
- **< 0.0**: Rastgeleden daha kötü uyum



## ✨ Cohen's Kappa Formülü

Cohen's Kappa şöyle hesaplanır:

Kappa = (Po - Pe) / (1 - Pe)

Burada:

- **pₒ (observed agreement)**: Gözlemlenen (gerçek) uyum oranıdır. Yani, iki değerlendiricinin aynı kararı verdiği örneklerin oranıdır.
- **pₑ (expected agreement)**: Rastgele aynı kararı verme olasılığıdır.

---

### 🔎 Formülün Yorumu

- Eğer değerlendiriciler tamamen aynı kararı verirse \( p_o = 1 \) olur, bu durumda \( \kappa = 1 \).
- Eğer uyum tamamen rastgele ise, \( p_o = p_e \), bu durumda \( \kappa = 0 \).
- Eğer \( p_o < p_e \), yani uyum rastgeleden daha kötüyse, \( \kappa < 0 \).

Bu formül sayesinde rastgele tahminin etkisi ortadan kaldırılır ve **gerçek uyum kalitesi** ölçülür.



## 👨‍⚖️ Değerlendirici Etiketleri Ne Anlama Gelir?

- **Birinci değerlendirici (true_labels):** Gerçek sınıflar (ground truth) veya ilk insan etiketi.
- **İkinci değerlendirici (pred_labels):** Modelin tahmini ya da başka bir insan etiketi.

Cohen's Kappa, iki değerlendiricinin aynı örneğe ne kadar benzer karar verdiğini ölçer. Örneğin bir doktor ve bir modelin teşhis uyumu, ya da iki doktorun teşhisi.


In [1]:

from sklearn.metrics import cohen_kappa_score


In [2]:

# Gerçek etiketler veya ilk değerlendiricinin etiketleri
true_labels = [1, 0, 2, 1, 0, 2, 2, 1]

# İkinci değerlendiricinin (veya modelin) tahminleri
pred_labels = [1, 0, 2, 0, 0, 2, 1, 1]

# Cohen's Kappa skorunu hesaplayalım
kappa = cohen_kappa_score(true_labels, pred_labels)

# Sonucu ekrana yazdıralım
print("Cohen's Kappa Skoru:", kappa)

# 📌 Açıklama:
# true_labels: Gerçek sınıflar (veya birinci değerlendirici)
# pred_labels: İkinci değerlendirici veya model tahminleri
# cohen_kappa_score(): İki etiket seti arasındaki kappa skorunu döndürür


Cohen's Kappa Skoru: 0.627906976744186



## 📊 Sonuç

Bu örnekte hesapladığımız Cohen's Kappa skoru, iki değerlendirici (veya model ile gerçek etiket) arasındaki **uyumun derecesini** gösterir.

---

## 💡 Skor Değerlendirmesi

Bu örnekte çıkan **0.6279** civarındaki skor, **iyi düzeyde bir uyuma** işaret eder. Genel olarak kappa skorları şöyle yorumlanabilir:

- **< 0.00**: Kötü uyum (rastgele tahminden daha kötü)
- **0.00 – 0.20**: Zayıf uyum
- **0.21 – 0.40**: Düşük uyum
- **0.41 – 0.60**: Orta düzey uyum
- **0.61 – 0.80**: İyi uyum
- **0.81 – 1.00**: Çok iyi (mükemmel) uyum

Bu durumda skorumuz **0.6279**, "iyi" kategorisine girer ve değerlendiriciler arasında anlamlı bir uyum olduğunu gösterir.

---

## ✅ Özet

- Cohen's Kappa, iki sınıflandırıcı arasındaki uyumu ölçmek için güçlü bir metriktir.
- Tesadüfi uyum ihtimalini göz önünde bulundurur.
- Özellikle etikete dayalı çalışmalarda tercih edilir.

---

🎁 **İpucu:** Dengesiz veri setleriniz varsa, Cohen's Kappa skorunun düşük çıkabileceğini ve bunun her zaman kötü performans anlamına gelmediğini unutmayın.
